### DISCUSSION TITLE
Smarty nocache condition

### DISCUSSION THREAD

**Discussion by madgreen1** [smarty-php/smarty#641] [Q&A] [answered]
Hi!
Back in version 3.1.18 I successfully used this syntax:
```
{nocache}{if (!empty($smarty.session.user_id))}{/nocache}
    {$cashed_var}
    <script>{literal}
        alert ("!");
    {/literal}</script>
{nocache}{/if}{/nocache}
```
It was very convenient to create non-cached conditions with the cached content inside a cached template. Now (in version 3.1.35) I get an error:
`
Syntax error in template "file:/file.tpl" on line 4 "{literal}" unclosed '{if}' tag in file "/Smarty/libs/sysplugins/smarty_internal_templatecompilerbase.php" at string #4
`

How can I resolve an issue with a non-cached condition when using the Smarty cache?
@wisskid can you help? A similar problem is described here https://stackoverflow.com/questions/7731332/need-trick-to-use-cached-value-in-nocache-section

**Comment by wisskid** [comment, marked as answer]
Seems rodneyrehm already answered it there accurately. Im afraid this isn’t possible and there won’t be a Smarty 3.2 in the near future.

Given a more elaborate use case we might be able to come up with another solution?

**Reply by madgreen1** [reply]
Thank you so much for your reply! I thought that in 10 years there was some solution for passing a cached variable to a non-cached area... In addition, we are already talking about working on Smarty 4 :)

The ideal solution (from the point of view of the frontend) would be
```
{if (some condition) nocache} // <- only it's not cached
    {$cashed_var}
{/if} // <- and this
```
Is it possible to implement this in version 4?

**Reply by wisskid** [reply]
I'm sure it's possible but I cannot promise it will be done. What is it that you are trying to achieve? There might be some other way to do it.

**Reply by madgreen1** [reply]
A fairly large site uses Smarty caching. The content of the pages is determined by user authorization (controls, various informers, counters - some of this data is selected from the database and cached) and you need to show it by condition.

For example, a feed of various objects, in which an authorized user can perform various actions on an object, but you need to pass the object type and other data:
````
{if (!empty($smarty.session.user_id))}
    <a data-type="{$item.type}" role="button" data-id="{$item.id}">{$item.data}</a>
{else}
    <a role="button">{$item.data}</a>
{/if}
````
In this case, {$item.type}, {$item.id}, {$item.data} - must be cached. In reality, templates are much more complex... There is data that an unauthorized user should not see in the page code in principle. And there are those that simply increase traffic unnecessarily (large json). Yes, we can transmit redundant data for all users, but sometimes overhead data can take up 25-30% of the total data volume.

An alternative solution would be a re-parsing plugin. But it seems that this is a wrong way, breaking the whole caching paradigm.

In fact, {nocache} is great, but it lacks the ability to transfer cached data to it.

**Reply by wisskid** [reply]
I see! At iWink, we have implemented this using a custom compiler function that creates a `{cache} .... {/cache}` block. It supports a cache_key attribute, or uses a hash of all attributes passed in to create a cache-id.

This allows you to write your code very explicit, like this:
```
{if (!empty($smarty.session.user_id))}
   {cache cache_key="auth"|cat:$item.type|cat:'-'|cat:$item.id}
    <a data-type="{$item.type}" role="button" data-id="{$item.id}">{$item.data}</a>
   {/cache}
{else}
  {cache cache_key="public"|cat:$item.type|cat:'-'|cat:$item.id}
    <a role="button">{$item.data}</a>
   {/cache}
{/if}
```

If you're interested, I can see if I can share a generalized version of this block and we could consider make it available by default in Smarty4.

**Reply by madgreen1** [reply]
The {cache} block looks like a great solution! It would be just great if this feature was implemented in version 4!

**Comment by madgreen1** [comment]
Hey @wisskid !
What about {cache} block in Smarty4? Is it not possible to implement yet?

**Reply by wisskid** [reply]
I'll look into it shortly!
